Use physical values for voxel bounds

jjhembd opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

This PR changes the way we handle the bounds of a voxel dataset. Previously, `VoxelPrimitive.prototype.minBounds` and `VoxelPrimitive.prototype.maxBounds` were defined in a "shape space", spanning nominal ranges such as [0..1], [-1.. 1], [-pi/2.. pi/2], or [-pi.. pi]. This was hard to document consistently, and inconsistent with the way bounds are defined in 3D Tiles.

After this PR, the `minBounds` and `maxBounds` are physical values, consistent with what is defined in the tileset.json.

## Issue number and link

Resolves https://github.com/CesiumGS/cesium/issues/12570.

## Testing plan

Load the [Voxels in 3D Tiles Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html?src=Voxels%20in%203D%20Tiles.html), and play with the clipping bounds. The extreme values of the sliders should represent the physical dimensions, and clipping should work normally.

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

ggetz commented on packages/engine/Source/Scene/VoxelCylinderShape.js:
```diff
@@ -59,40 +59,16 @@ function VoxelCylinderShape() {
-
-  /**
-   * @type {number}
+   * @type {Cartesian3}
```
Even though it's private, let's document what each component of the cartesian corresponds to– angle, height, range.

ggetz commented on packages/engine/Source/Scene/VoxelCylinderShape.js:
```diff
@@ -163,84 +143,45 @@ VoxelCylinderShape.prototype.update = function (
+  clipMinBounds.y = CesiumMath.negativePiToPi(clipMinBounds.y);
+  clipMaxBounds.y = CesiumMath.negativePiToPi(clipMaxBounds.y);
+
+  // TODO: what if a min angle is > a max angle? (crossing the 180th meridian)
```
Discussed offline, but since this is an edge case, we can document this as a known issue rather than fix this here. Please open a new issue, if you haven't done so already. 🙂

jjhembd replied:
False alarm: I tested this, and clipping across the antimeridian actually works fine!
I removed the TODO comments.

ggetz reviewed (commented):
Thanks @jjhembd! Besides the know TODO's, code is looking good so far. Let me know when you'd like a final review and merge.

jjhembd commented:
Thanks for the review @ggetz! I think I addressed all your feedback and this is ready for another look.
I also tested it on several more datasets, both BOX and CYLINDER. Everything works as expected, including vertical exaggeration.

ggetz commented:
@jjhembd Testing this out, everything is working great in the "Voxels" example. It's much more intuitive controlling the bounds and clipping bounds with this change.

I tried the "Voxels in 3D Tiles" example and modifying the clipping bounds all works as expected. However, for the "Bounds", adjusting any of the components seems to visually set that component to `0`. This applies to both Box and Cylinder shapes. Is that expected?

jjhembd commented:
> However, for the "Bounds", adjusting any of the components seems to visually set that component to `0`.

@ggetz The slider ranges for the "Bounds" are currently not adjusted in `VoxelInspectorViewModel`. I just now tried making them update when a new `VoxelPrimitive` is tracked, but it's creating a funny feedback loop which might take a little while to figure out (we read the min/maxBounds to set the clippingBounds, etc).

Long term I think we should make the bounds fixed. Adjustable bounds was implemented for debugging (I assume), but for a finished app, I think adjusting bounds is really a task for a model editor rather than a renderer.

Would it make sense to remove the "Bounds" tracker from the `VoxelInspector` for now? Then we could have a followup later to make the bounds read-only in the `VoxelPrimitive` API itself, perhaps after we have had a chance to discuss further.

ggetz commented:
> Long term I think we should make the bounds fixed. Adjustable bounds was implemented for debugging (I assume), but for a finished app, I think adjusting bounds is really a task for a model editor rather than a renderer.

> Would it make sense to remove the "Bounds" tracker from the VoxelInspector for now? Then we could have a followup later to make the bounds read-only in the VoxelPrimitive API itself, perhaps after we have had a chance to discuss further.

This all tracks. Do you plan on removing those properties from the voxel inspector in this PR, or a follow-up?

lrjcx commented:
@ggetz [Sandcastle](https://sandcastle.cesium.com/#c=rVhbb9vIFf4rA78sGTMjkZJsy7LTpvbuIkBugN31iwHviBxJg/DWmaFkeWGgL0W3L+nTYlugaF8WWKAvLdBi/1G8yL/oOTOkREpUmhQNEHk4c+7nm3MOGXNN5oIvuCSn5LPPRtdpmKVKE5EK/VW1z9QyDYnjktMn5JvrlBBLc/7qBRxGWVgkPNV0yvXnMcflr5bPIud6L+RKFMlZlmomUi6v99wRMq/UpXxBzgwNtaockOhZDYSoWbb4UmZFGj3VSabyGZf8mExYrLhnKSKe69klV/rpFBQofcmlhMUx0bKoaMZM8edsyeVxpepZwqZcLs0mncgseS2zuYi4fIpeOpaNVNRPZfilUC9YfsHlnMuSuWIx/L+W8YqLkOu9mda5Ou50FHAIiAFlMpwKlaUxRIGGWdJBoc8uOhJMX1F1rjIZRzelgs5K4/VeJdu1C7fyfpmyRITr4DwX05kW6fQLsOmiSOuxuq9Hnt7YxFyJCFIGT5JHQq/SRJVexpxGQuUxW0KarvfSLOXXe3URKuTgyzTOxpzylI3jlXZgMHq3qVWRUkzqbookyz5AMmNRtlAfkP9muQ5GJQah8H/2XVuYVRjA+7FgQleIAZFMc5PNEpAWWCWsJf9NAXm/Ahr5gqk3DbSWh19xqfnty0wm4GmNoJnGkCVcMjqJl5dZJT0CbpEyLbJ0BfgzBtKUYGnPwPWcTyXnyvH9Q3oQ9PqHPY/0u/TAPxgeesTvdrsVwjIp4DKXwr6pYDjjLII8H5M+7Q2HfX8IIgaD3sHhkVeR5EKHs2PSpSBzOOj3eyD0oNcd9FcUMovjY3JAg6OePzgKDgd+cACG2OP7CuGFLJX31u7jQnF9KRKeFdqpFSVCWBQhDB0bpnuPoGJ8uDeMtmqVRLvL2kSyKVaxCwCcSe/XldVFKiaQFKJYksdcBufAEWfykt/qQlqYtZOVlapJJlIy52FA5jfanpxlmYwwe1wZoopwEmdME5lGjqHP3XUy1scTvBiShdpRInWiTDu5Z+Q7fkB9yLNHDgN61BsGruuSR6QPIe/RwB3VRUkOdqRkstq8b7fDrhYfYcgCdSGo6P+giU8hBzkv3fYgYpq8bFc62t5jmGLNUientx7J6dIl+4DJoIV0DKQWjEPSsXvOy6bB6JaCAqVnSvPc6dKBB8IGvgcYUBiQTOImqGAgYoxuj8ljwnAR83SqZ2jH0v2UMIDfPRIKGcbrGFiDlbjj1TriIVt6JTHisb4OVlRCaQsIkmSF4rUwdjrk4e2P77796d3v/vDw5+/Iw3ffPvz1L9tRiiEElSvgpxED3jkoGhFFXYwwmgYRCGhLnOPgU0T02kSEICFhtxDrrg/xzbOF0yIQ5bmeFQT4o30Xs9XFZAy6rWL9tdy1YPjbpVAffQrmQJG0xqVw8AhqC10LbTc1WMvsg4CGrY8btpqAGXh2h+bGUFRayYfDVqPVdixWdyZHV+lGSKq9Lh165MCt68CTRtXBfwZCoMQCG38fGf/NfhPJFSneBoQfGNXvo+zA/qKGI9S9TjX8rgQ3ZBn44oULgcQI3riK+6eYsN1nwe4z1XpU3UOT8K7/oaoEY8zqMrZep11VyRADdw6OrbDaltUiBQqfdsGYEi7IZQP4KdUjExHAA9LVapzAFooxvktuoF4n/GWRjKHfdUy53sAb2N4nZsw6K/Nc9iyn3gG91l7mbkehmIOEaXzzBbRaQwmVERSjKTjc5JnU9G5hktFERjGnt+TR6QblJuuiJaqmBZfuQjRZ+OYcd5yVH7Bv5tnzWq/e4ZDbEpsifZ2pUn6SyXxmw2sT+QtDZHRYQoFzzdUZXkC4e8etxxBpyEih8N2I3t0uLfEu1Ybl8zOyjg3eVXP0QZarioUv+WV2JVIYsmu+Og3pTe0Nn6vgv5J6lk0ly2cCb69xrKGOoiePm1t37b7B3Uqw7RpahEiDIEk+Hgtb1QW7agXlslwNTZ3CH98d7WYI1hw9Q4w/Q3eHmlJFIm6dOpNR0w2wtBvF5ZnZNoPFEdRlAPvSNhd7E4Ky+pqb646wb9tL/O6Hv7//07+ge//89h9lJ//hx3f//CN5+P6n99//+/1v//bz299v2gcjKikHOWFK/Aj+nsDlx8X+/kbNWJd4rK92JCnmnu05MBEKsC0wY57pXkce9DPsltB6TYC6Xi3k+0R49YCa57UUWjptpjXrugfZbiblfsuf5tCHwcMmm9ClZ8uGebh1R+2DDfbRRs2FGoUcrjHgqG/b5Khd61ji26+Jot/SFferNu03p4CVCmy+HvxfN2M2VmZ+ZlXLxr6KD0OK8/v2RPDRmiDB5UATrCeY/YbCXptCaN4wqNXJqhN/hyUOmlKGZpc1/ZUxNeftbFC/E4EBUc8e7hhToBTw27w9iWjmgLYWQlPvt7raJ3SD9hZ+XvabhgbanDrEBCZfQ/dkRbd17+Att+yT1sR1F27eB8JjmPP+C7PtM/YlwbfBNlub+5QxVlLs1/s+ICAoB+RtlpYLWlt8PVq/YueZ0q9lFnKlLjSb8ubHwNcbp9XHjeaL+fHGc8sXEvtdaFOZoiyKnM3d6jPB+sOn4472vL0T80noSeXML0WCbYUUMnYo7WgOr/iIgs64CN9wTUO1AsRJp856Eok5EdFpy2dREsZMKTiZFHF8Aa8s13tPTjpAv8UK4MJvL6/mXMZsiWQz/8lzu0kpPenAYzunzrJ4zOSG5P8A)   That's the current address, thank you。 When the camera moves to the sky, it will report an error.

ggetz commented:
@lrjcx Thanks for the example! Could you please let us know how that is related to this PR? If it's not, could you please open a new issue, or ask a question on the [Cesium Forum](https://community.cesium.com/)?
